packages/devtools_app/lib/src/charts/treemap.dart

```diff
@@ -0,0 +1,493 @@
+  // |  |       |  |
+  // |  |       |  |
+  // |  ---------  |
+  // ---------------
+  final int levelsVisible;
+
+  final double width;
+  final double height;
+
+  final VoidCallback onTap;
+
+  @override
+  _TreemapState createState() => _TreemapState();
+}
+
+enum PivotType { pivotByMiddle, pivotBySize }
+
+class _TreemapState extends State<Treemap> {
+  PivotType pivotType = PivotType.pivotBySize;
+  static const double minHeightToDisplayText = 20.0;
+  bool shouldDisplayText;
+
+  TreemapNode rootNode;
+
+  @override
+  void initState() {
+    super.initState();
+    rootNode = widget.rootNode;
+    shouldDisplayText = widget.height > minHeightToDisplayText;
+  }
+
+  void cellOnTap(String name) {
+    final TreemapNode child = rootNode.childrenMap[name];
+    if (child != null) {
+      setState(() {
+        rootNode = child;
+      });
+    }
+  }
+
+  // Computes the total size of a given list of treemap nodes.
+  int computeSize(List<TreemapNode> children, int start, int end) {
+    int sum = 0;
+    for (int i = start; i <= end; i++) {
+      sum += children[i].byteSize;
+    }
+    return sum;
+  }
+
+  int computePivot2(List<TreemapNode> children) {
+    switch (pivotType) {
+      case PivotType.pivotByMiddle:
+        {
+          return (children.length / 2).floor();
+        }
+      case PivotType.pivotBySize:
+        {
+          int pivotIndex = -1;
+          double maxSize = double.negativeInfinity;
+          for (int i = 0; i < children.length; i++) {
+            if (children[i].byteSize > maxSize) {
+              maxSize = children[i].byteSize.toDouble();
+              pivotIndex = i;
+            }
+          }
+          return pivotIndex;
+        }
+    }
+    return -1;
+  }
+
+  // Divides a given list of treemap nodes into four parts:
+  // L1, P, L2, L3
+
+  // P (pivot) is the treemap node chosen to be the pivot based on the pivot type.
+
+  // L1 inclues all treemap nodes before the pivot treemap nodes.
+
+  // L2 and L3 combined include all treemap nodes after the pivot treemap node.
+  // A combination of elements are put into L2 and L3 so that
+  // the aspect ratio of PD is as low as it can be.
+
+  // Example layout:
+  // ----------------------------
+  // |      |  PD  |            |
+  // |      |      |            |
+  // |  L1  |------|    L3      |
+  // |      |  L2  |            |
+  // |      |      |            |
+  // ----------------------------
+  List<Cell> buildTreemap(
+    List<TreemapNode> children,
+    double width,
+    double height,
+    double x,
+    double y,
+  ) {
+    final bool isHorizontalRectangle = width > height;
+
+    final int totalSize = computeSize(children, 0, children.length - 1);
+
+    if (children.isEmpty) {
+      return [];
+    }
+
+    // Sort list of children, descending in size.
+    children.sort((a, b) => b.byteSize.compareTo(a.byteSize));
+
+    if (children.length == 1 || children.length == 2) {
+      final List<Cell> positionedChildren = [];
```

Comment:
instead push the type to the RHS and write `final positionedChildren = <Cell>[];`

Reply:
Done